ADS-3608: Add scope in Dynamic bundle discounts documentation (#80)
Nosto/wiki-docs: features/dynamic-bundle-discounts.md

Secret key, on the other hand, can be retrieved from "Dynamic Bundle Key" field in Nosto Admin Settings > Platform page. This value replaces GET\_FROM\_NOSTO placeholder in the Shopify line item script.

## Scope
In scope - Apply discounts when all the products in a bundle are added to cart. 
Not in scope - Selecting only specific products from a bundle, for discounts.

> ## Will not work If

Reviewer: 'will not work if' doesn't read well. I think it's better to

1. Add a sentence to the introduction to say that all products need to be added to the cart in order for the discount to be applied.
2. Add a warning to the Nosto Bundle Script section to warn that we can only provide support for the script as-is and we can't provide any support for scripts that have been modified.

Secret key, on the other hand, can be retrieved from "Dynamic Bundle Key" field in Nosto Admin Settings > Platform page. This value replaces GET\_FROM\_NOSTO placeholder in the Shopify line item script.

## Scope
Scope of this initial release is to support discounts on bundled product offerings. 

### Bundle Discount Example

> Let's say, in PDP for Product#1, we display a bundle of products Product#2, Product#3 & Product#4 along with the only action button "Add to cart". 

Reviewer 2: Just clarifying out internal jargon
```suggestion
In the Product Detail Page (PDP) for Product#1, we display a bundle of products Product#2, Product#3 & Product#4 along with the only action button "Add to cart". 
```

### Supported Scenarios
The logic works only under the following scenarios:
_Note: This section is subjected to change in future_

#### When all the products of a bundle offering are added to the cart
     
Consider the example in the [Bundle Discount Example](#bundle-discount-example) section above
Upon clicking "Add to cart" button, Nosto script adds all the products of bundle Product#2, Product#3 & Product#4 to cart and applies the pre-configured discount

> #### When all the products of a bundle offering are added to the cart alogn with PDP product

Reviewer 2: ```suggestion
#### When all the products of a bundle offering are added to the cart along with PDP product
```

Recommendation templates for bundle offerings can be configured to include currently displayed product for discount when a bundle is added to car, using the `discountCurrentProduct` discussed under section [Nosto Dynamic Bundle template](#nosto-dynamic-bundle-template) above.

Consider the example in the [Bundle Discount Example](#bundle-discount-example) section above and assuming the `discountCurrentProduct` variable in the recommendation template is set to `true`,
Upon clicking "Add to cart" button, Nosto script adds all the products of bundle Product#2, Product#3, Product#4 and **Product#1** (currently displayed product), to cart and applies the pre-configured discount

### Scenarios not supported

#### When a partial bundle is added to cart

When we offer the option of selecting only specific products from a bundle, say with a checkbox or option button, in the recommendation template, specific products can be selected or deselected from a bundle and then added to cart. In this situation, the logic will not apply discount on any product.   

> Consider the example in the [Bundle Discount Example](#bundle-discount-example) section above and assuming each product in the bundle has an associated checkbox, for selecting or deselecting a product

Reviewer: This is quite a verbose description of the issue. Isn't it simply that all products from the bundle need to be added to cart for the discount to be applied?

Author: IMO even that is not required as the title itself conveys this message "Dynamic Bundle discounts". But the questions posted by Pouria still comes along

Product#3 in the bundle is deselected using the checkbox and "Add to cart" button is clicked. In this scenario, only Product#2 and Product#4 gets added to cart with no discounts. 

**In order to overcome this issue, the template shouldn't include any option for selecting or deselecting products**

## Will not work If

  1. Nosto script for applying discount, provided [here](#authentication-script) and [here](#nosto-bundle-script), is not used as it is (except for replacing the SECRET_KEY)

  2. SECRET_KEY is not configured correctly in [Authentication Script](#authentication-script)

> 

Reviewer: This section just needs to be added to the Known Issues section, doesn't it? Why the need for a new section with additional description of the feature?

Author: I don't think so. `Known Issues` section contains issue that's not created by us. But these issues are created when we add flavours to our recommendation templates

end

Output.cart = Input.cart
```

## Troubleshooting

> ### Issue #1: When additional content displayed in cart page

Reviewer: ```suggestion
### Issue #1: Additional content displayed on cart page
```

![Properties Exposed](https://user-images.githubusercontent.com/82023195/168611801-666c24bc-c7f9-4edd-bdf5-2be32798391c.png)

### Cause

> Nosto implementation uses a private property field for sharing discount information with Shopify. In newer theme versions, Shopify automatically hides these fields while displaying products in cart page. In older version of themes, these fields gets exposed as shown below.

Reviewer: > private property field
Do you mean a hidden form value? e.g `<input type="hidden">`?

Author: Rather it's something that's not supposed to be exposed. "Technically" we can assume it's a hidden field

### Solution

In such a case, make the following changes to the template liquid file that displays the cart information (usually cart.liquid or cart-template.liquid).

```html
{% raw %}
{% for p in item.properties %}
  {% assign first_character_in_key = p.first | truncate: 1, '' %} <== new line
  {% unless p.last == blank or first_character_in_key == '_' %} <== new line
  <small>
    {{ p.first }}:

    {% comment %}
      Check if there was an uploaded file associated
    {% endcomment %}
    {% if p.last contains '/uploads/' %}
      <a href="{{ p.last }}">{{ p.last | split: '/' | last }}</a>
    {% else %}
      {{ p.last }}
    {% endif %}
  </small><br />
  {% endunless %} <== new line
{% endfor %}
{% endraw %}
```

> ### Issue #2: When discount is not applied

Reviewer: ```suggestion
### Issue #2: Discount is not applied
```

### Cause

> 1. Nosto provided scripts are not used as-is. Nosto can't provide support in this case.

Reviewer: ```suggestion
1. Nosto cannot provide support for the line item script if any customisations are made to it.
```

> 2. Incorrect value used for SECRET_KEY. 

Reviewer: ```suggestion
2. The `SECRET_KEY` in the line item script is incorrect.
```

Author: I think the new sentence is little confusing. It suggests that the `SECRET_KEY` is incorrect rather than it's value

3. Issue with Bundle script configuration.

> 4. Not all the products from a bundle is added to cart.

Reviewer: ```suggestion
4. Not all the products from a bundle are added to cart.
```
